contrib/babelfishpg_tsql/src/pl_handler.c

```diff
@@ -6391,6 +6476,796 @@ bbf_ExecDropStmt(DropStmt *stmt)
+		partition_names = lappend(partition_names, get_rel_name(inhrelid));
+	}
+	systable_endscan(scan);
+	table_close(relation, AccessShareLock);
+
+	/*
+	 * We prepare the following query to rename the partitions of partitioned table.
+	 * This will be executed using standard_ProcessUtility().
+	 */
+	initStringInfo(&query);
+	appendStringInfo(&query, "ALTER TABLE \"%s\".dummy RENAME TO dummy;", physical_schema_name);
+
+	old_dialect = GetConfigOption("babelfishpg_tsql.sql_dialect", true, true);
+
+	/* We need to change the dialect to postgres to parse the RENAME statement. */
+	PG_TRY();
+	{
+		set_config_option("babelfishpg_tsql.sql_dialect", "postgres", GUC_CONTEXT_CONFIG,
+					PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+
+		parsetree = raw_parser(query.data, RAW_PARSE_DEFAULT);
+		rename_partition_stmt = (RenameStmt *) parsetree_nth_stmt(parsetree, 0);
+	}
+	PG_FINALLY();
+	{
+		/* Reset dialect back to original value. */
+		set_config_option("babelfishpg_tsql.sql_dialect", old_dialect, GUC_CONTEXT_CONFIG,
+					PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+	}
+	PG_END_TRY();
+
+	if (!rename_partition_stmt) /* Sanity Check. */
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("Failed to rename partitions of relation \"%s\".", table_name)));
+	}
+
+	/* Need to make a wrapper PlannedStmt. */
+	wrapper = makeNode(PlannedStmt);
+	wrapper->commandType = CMD_UTILITY;
+	wrapper->canSetTag = false;
+	wrapper->utilityStmt = (Node *) rename_partition_stmt;
+	wrapper->stmt_location = 0;
+	wrapper->stmt_len = 0;
+
+	/* Construct hash based on new name. */
+	new_hash = construct_unique_hash(stmt->newname);
+	
+	for (int i = 0; i < list_length(partition_names); i++)
+	{
+		partition_name = list_nth(partition_names, i);
+
+		/*
+		 * Generate new partition name by replacing the hash portion
+		 * of existing partition name with new hash value.
+		 */
+		new_partition_name = pstrdup(partition_name);
+		memcpy(new_partition_name, new_hash, MD5_HASH_LEN);
+		
+		/* Set the names in RENAME statment. */
+		rename_partition_stmt->relation->relname = partition_name;
+		rename_partition_stmt->newname = new_partition_name;
+
+		/* Execute the rename statment. */
+		standard_ProcessUtility(wrapper,
+					"(RENAME PARTITION)",
+					false,
+					PROCESS_UTILITY_SUBCOMMAND,
+					NULL,
+					NULL,
+					None_Receiver,
+					NULL);
+
+		CommandCounterIncrement();
+
+		pfree(partition_name);
+		pfree(new_partition_name);
+	}
+
+	/* Free the allocated memory. */
+	if (partition_names)
+		list_free(partition_names);
+	pfree(logical_schema_name);
+	pfree(new_hash);
+	pfree(query.data);
+}
+
+/*
+ * For Babelfish partitioned tables, non-superusers should not be permitted
+ * to attach or detach partitions from the partitioned table, and they
+ * should also be restricted from modifying the partitions from both
+ * TSQL as well as PG endpoint.
+ * 
+ * NOTE: We are only blocking operation on Babelfish partitioned tables i.e.
```

Comment:
Lets speak to DBEs and get the CREATE PARTITIONED TABLES to be blocked from psql